activerecord/lib/active_record/associations/preloader/through_association.rb

```diff
@@ -25,18 +24,18 @@ def run(preloader)
               owner.association(through_reflection.name).reset if through_scope
             end
             result = through_records.flat_map do |record|
-              association = record.association(source_reflection.name)
-              target = association.target
-              association.reset if preload_scope
-              target
+              record.association(source_reflection.name).target
             end
             result.compact!
-            if reflection_scope
-              result.sort_by! { |rhs| preload_index[rhs] } if reflection_scope.order_values.any?
-              result.uniq! if reflection_scope.distinct_value
-            end
+            result.sort_by! { |rhs| preload_index[rhs] } if scope.order_values.any?
+            result.uniq! if scope.distinct_value
             associate_records_to_owner(owner, result)
           end
+          unless scope.empty_scope?
+            middle_records.each do |owner|
+              owner.association(source_reflection.name).reset
```

Comment:
For some reason I have a collection where some of the `middle_records` are `nil` and this explode. I'll try to reproduce but does someone know if that can be the case?

Reply:
hmm... somehow preloaded records have `nil`, at least that is skipped in `grouped_records` for now.
(I removed the skip in the past, but as far as I remember tests aren't passed.)

https://github.com/rails/rails/blob/bd4eff2f99ffa5868574c497bc63f320cbda3083/activerecord/lib/active_record/associations/preloader.rb#L152-L155

Reply:
Got a reproduction script:

```ruby
# frozen_string_literal: true

require "bundler/setup"

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end

  create_table :foos, force: true do |t|
    t.integer :post_id
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
  end
end

class Post < ActiveRecord::Base
  has_many :comments
  has_one :foo
end

class Comment < ActiveRecord::Base
  belongs_to :post

  has_one :foo, through: :post
end

class Foo < ActiveRecord::Base
  default_scope { order(:id) }

  belongs_to :post
end

class BugTest < Minitest::Test
  def test_association_stuff
    Comment.create!
    Foo.create!

    assert_equal 1, Comment.preload(:post, :foo).to_a
  end
end
```

I'll write a test and commit the fix. Not sure if we should fix the root cause of having `nil`s at that point or just skip if there are `nil`s. WDYT?

Reply:
Ideally we'd like to find the root cause. But for now just skip `nil` record is also fine to me since we already have that skip in `grouped_records` for a long time.

Reply:
Fixed in 572dcdd7e8